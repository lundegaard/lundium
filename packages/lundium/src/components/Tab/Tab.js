import React from 'react';
import PropTypes from 'prop-types';
import { isNil } from 'ramda';
import { cx } from 'ramda-extension';

import Box from '../Box';
import Link from '../Link';

const Tab = ({
	className,
	label,
	isActive,
	isDone,
	isDisabled,
	isInvalid,
	stepNumber,
	href,
	...otherProps
}) => {
	const WrapperComponent = isDisabled || isNil(href) ? 'span' : Link;

	return (
		<WrapperComponent
			className={cx(
				'tab-link',
				{
					['tab-link--active']: isActive,
					['tab-link--done']: isDone,
					['tab-link--disabled']: isDisabled,
					['tab-link--invalid']: isInvalid,
					['tab-link--wizard']: !isNil(stepNumber),
				},
				className,
			)}
			{...(isDisabled ? {} : { href })}
			{...otherProps}
		>
			{!isNil(stepNumber) && (
				<span className="tab-link__number">{stepNumber}</span>
			)}
			{label && (
				<Box as="span" className="tab-link__label">
					{label}
				</Box>
			)}
		</WrapperComponent>
	);
};

Tab.propTypes = {
	/** Class name for root. */
	className: PropTypes.string,
	href: PropTypes.string,
	/** If `true`, tab link is in active state. */
	isActive: PropTypes.bool,
	/** If `true`, tab link is disabled for any pointer events. */
	isDisabled: PropTypes.bool,
	/** If `true`, tab link is in done state, applied in step wizard. */
	isDone: PropTypes.bool,
	isInvalid: PropTypes.bool,
	/** Label for tab link. */
	label: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
	/** Step number presents numbered position in step wizard. */
	stepNumber: PropTypes.number,
};

export default Tab;
