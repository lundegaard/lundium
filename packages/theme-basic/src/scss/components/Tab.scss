@import '../core';

/*
 -  Tab Link
 -  (<a href="#*" class="tab-link" />)
 */


.tab-link {
	flex-grow: 0;
	flex-shrink: 0;
	position: relative;
	padding-bottom: $spacer;
	display: inline-flex;
	align-items: center;
	color: $tab-link-color;
	font-size: $tab-link-font-size-mobile;
	font-weight: 400;
	white-space: nowrap;
	cursor: default;
	overflow: hidden;
	transition:
		flex ( $transition-speed * .5 ) $transition-easing,
		font-weight ( $transition-speed * .5 ) $transition-easing,
		color $transition-speed $transition-easing;

	// Media queries
	@include media-breakpoint-up(md) {
		font-size: $tab-link-font-size;
	}

	&:not(.tab-link--active):not(.tab-link--disabled) {
		cursor: pointer;
	}

	&:hover:not(.tab-link--disabled) {
		text-decoration: none;
		color: $tab-link-hover-color;

		&.tab-link--active {
			color: $tab-link-color;
		}
	}

	&:not(.tab-link--wizard)::after {
		content: '';
		position: absolute;
		inset-block-end: 0;
		inset-inline-start: 0;
		height: 2px;
		width: 100%;
		background-color: $tab-link-active-bg-color;
		transform: translateY(100%);
		transition: transform $transition-speed $transition-easing;
	}

	&--invalid {
		color: $danger;
	}

	&--disabled {
		opacity: .3;
		cursor: not-allowed;

		&:hover {
			color: $body-color;
			text-decoration: none;
		}
	}

	&--active {
		font-weight: 800;

		.tab-link__number {
			border-color: $tab-link-active-bg-color;
			background-color: $tab-link-active-bg-color;
			color: set-color-by-bg($tab-link-active-bg-color);
		}

		&:not(.tab-link--wizard)::after {
			transform: translateY(0);
		}
	}

	&--done {
		.tab-link__number {
			border-color: $tab-link-done-bg-color;
			background-color: $tab-link-done-bg-color;
			color: set-color-by-bg($tab-link-done-bg-color);
			font-weight: 800;
		}
	}

	&--done,
	&--disabled {
		.tab-link__number + .tab-link__label {
			// Media queries
			@include media-breakpoint-down(xs) {
				display: none;
			}
		}
	}

	& + & {
		margin-inline-start: ( $grid-gutter-width * .5 );

		// Media queries
		@include media-breakpoint-up(md) {
			margin-inline-start: $grid-gutter-width;
		}
	}

	&__number {
		width: $tab-link-number-size-mobile;
		height: $tab-link-number-size-mobile;
		border-radius: $tab-link-number-size-mobile;
		border: $tab-link-number-border;
		display: inline-flex;
		justify-content: center;
		align-items: center;
		font-size: $tab-link-number-font-size;
		line-height: 1;
		color: $tab-link-number-color;

		// Media queries
		@include media-breakpoint-up(md) {
			width: $tab-link-number-size;
			height: $tab-link-number-size;
			border-radius: $tab-link-number-size;
		}

		& + .tab-link__label {
			margin-inline-start: ( $spacer * .5 );

			// Media queries
			@include media-breakpoint-up(md) {
				margin-inline-start: ( $spacer * .75 );
			}
		}
	}
}