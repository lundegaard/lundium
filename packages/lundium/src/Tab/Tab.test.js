import React from 'react';
import { shallow } from 'enzyme';

import Tab from './Tab';

describe('<Tab />', () => {
	const requiredProps = {
		label: 'tab label',
	};

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<Tab {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
