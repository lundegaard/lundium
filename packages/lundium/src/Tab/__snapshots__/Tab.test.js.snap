// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tab /> matches snapshot when shallowly rendered 1`] = `
<span
  className="tab-link"
>
  <forwardRef(Box)
    as="span"
    className="tab-link__label"
  >
    tab label
  </forwardRef(Box)>
</span>
`;
